modelzoo/admin.py

```diff
@@ -0,0 +1,475 @@
+            if self.password == password and password != "":
+                if self.token == token and token != "":
+                    return self
+                else:
+                    u = UserAuth(
+                        conn,
+                        username=username,
+                        password=password,
+                        token=token,
+                        admin=False,
+                    )
+                    u.parent = self
+                    return u
+        u = UserAuth(conn, username=username, password=password, token=token)
+        return u
+
+    @property
+    def admin(self):
+        if self.adminF:
+            return self.admin_token
+        if self.parent is None:
+            return self.token
+        return self.parent.admin
+
+
+class ModelZooConnection(object):
+    def __init__(
+        self,
+        address: typing.Optional[str] = "grpc.modelzoo.live",
+        token: typing.Optional[str] = "",
+        username: typing.Optional[str] = "",
+        password: typing.Optional[str] = "",
+    ):
+        self.conn = None
+        self.address = address
+        self.conn_error = ModelZooConnectionException(
+            "Must be connected to ModelZoo instance before issuing requests. Please call connect()."
+        )
+        try:
+            self.auth = UserAuth(
+                self, username=username, password=password, token=token
+            )
+        except AuthenticationException as e:
+            self.auth = None
+
+    def connect(self, address: typing.Optional[str] = "grpc.modelzoo.live"):
+        self.address = self.address if address is None else address
+        channel = grpc.insecure_channel(self.address)
+        self.conn = ModelStub(channel)
+
+    def authenticate(
+        self,
+        username: typing.Optional[str] = "",
+        password: typing.Optional[str] = "",
+        token: typing.Optional[str] = "",
+    ) -> None:
+        try:
+            self.auth = UserAuth(
+                self, username=username, password=password, token=token
+            )
+        except InsufficientCredentialsException:
+            raise ModelZooConnectionException(
+                "At least one of [(Username, Password), Token] are required."
+            )
+        except InvalidTokenException:
+            raise ModelZooConnectionException("The token provided is not valid.")
+
+    def _authenticate(self, username: str, password: str) -> str:
+        if self.conn is None:
+            raise self.conn_error
+        resp = self.conn.Authenticate(
+            AuthenticationRequest(username=username, password=password)
+        )
+        if resp.success:
+            return resp.token
+        else:
+            raise ModelZooConnectionException(resp.msg)
+
+    def get_token(self):
+        if self.auth is None:
+            raise ModelZooConnectionException(
+                "Please authenticate before calling this method."
```

Comment:
Please call `conn.authenticate(username, password)` or `conn.authenticate(token)` before calling this method.

Reply:
Generally i like error message to be actionable